### PR TITLE
[Merged by Bors] - mesh: refactor Updates handling 

### PR DESCRIPTION
**Pull request by dshulyak**
still able to reproduce the bug below.  

> 2023-10-02T15:28:14.002+0200 WARN fd68b.sync mesh failed to process layer from sync {"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "sync", "sessionId": "29422935-68d6-47d1-87a8-02293aa181f3", "layer_id": 23104, "errmsg": "requested layer 8063 is before evicted 13102", "name": "sync"}

in the change i removed Results method from tortoise, and instead keeping Updates until mesh applies block and notifies tortoise that the block was applied, only then tortoise is free to evict results from memory. 

OnApplied has layer and opinion in the signature. Opinion is a unique representation of consensus state. If it changed while blocks were applied it means that history changed and we should not evict pending results.

### CODE REVIEW

File: syncer/syncer_test.go [hunk 1]
```diff
@@ -189,10 +199,14 @@ func TestSynchronize_OnlyOneSynchronize(t *testing.T) {
...
 	for lid := from; lid <= to; lid++ {
-		require.NoError(t, certificates.Add(ts.cdb, lid, &types.Certificate{BlockID: types.EmptyBlockID}))
+		require.NoError(
+			t,
+			certificates.Add(ts.cdb, lid, &types.Certificate{BlockID: types.EmptyBlockID}),
+		)
 		ts.mLyrPatrol.EXPECT().IsHareInCharge(lid)
 		ts.mDataFetcher.EXPECT().PollLayerOpinions(gomock.Any(), lid, false, gomock.Any())
 		ts.mTortoise.EXPECT().TallyVotes(gomock.Any(), lid)
+		ts.mTortoise.EXPECT().OnApplied(gomock.Any(), gomock.Any())
```
Comment (fasmat) [inline syncer/syncer_test.go]: Would it make sense to specify which parameters were used for the call?

```suggestion
		ts.mTortoise.EXPECT().OnApplied(lid, gomock.Any())
```

File: syncer/state_syncer_test.go [hunk 2]
```diff
@@ -410,11 +472,17 @@ func TestProcessLayers_NoHashResolutionForNewlySyncedNode(t *testing.T) {
 		ts.msh.SetZeroBlockLayer(context.Background(), lid)
 		ts.mTortoise.EXPECT().OnHareOutput(lid, types.EmptyBlockID)
 		ts.mTortoise.EXPECT().TallyVotes(gomock.Any(), lid)
-		ts.mTortoise.EXPECT().Updates().Return(fixture.RLayers(fixture.ROpinion(lid, types.RandomHash())))
+		ts.mTortoise.EXPECT().
+			Updates().
+			Return(fixture.RLayers(fixture.ROpinion(lid, types.RandomHash())))
+		ts.mTortoise.EXPECT().OnApplied(gomock.Any(), gomock.Any()).AnyTimes()
```
Comment (fasmat) [inline syncer/state_syncer_test.go]: Same here:

```suggestion
		ts.mTortoise.EXPECT().OnApplied(lid, gomock.Any()).AnyTimes()
```

File: syncer/state_syncer_test.go [hunk 3]
```diff
@@ -258,9 +288,12 @@ func TestProcessLayers_HareTakesTooLong(t *testing.T) {
 		}
 		peers := test.GeneratePeerIDs(3)
 		ts.mDataFetcher.EXPECT().SelectBest(gomock.Any()).Return(peers)
-		ts.mDataFetcher.EXPECT().PollLayerOpinions(gomock.Any(), lid, gomock.Any(), peers).Return(nil, nil, nil)
+		ts.mDataFetcher.EXPECT().
+			PollLayerOpinions(gomock.Any(), lid, gomock.Any(), peers).
+			Return(nil, nil, nil)
 		ts.mTortoise.EXPECT().TallyVotes(gomock.Any(), lid)
 		ts.mTortoise.EXPECT().Updates().Return(fixture.RLayers(fixture.RLayer(lid)))
+		ts.mTortoise.EXPECT().OnApplied(gomock.Any(), gomock.Any()).AnyTimes()
```
Comment (fasmat) [inline syncer/state_syncer_test.go]:
```suggestion
		ts.mTortoise.EXPECT().OnApplied(lid, gomock.Any()).AnyTimes()
```

**Review by fasmat — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**